src/Adapter/Order/OrderProductUpdater.php

```diff
@@ -0,0 +1,519 @@
+                            $orderDetail->product_attribute_id,
+                            new Warehouse($movement['id_warehouse']),
+                            $quantityToReinject,
+                            null,
+                            $movement['price_te']
+                        );
+                    }
+                } else {
+                    $manager->addProduct(
+                        $orderDetail->product_id,
+                        $orderDetail->product_attribute_id,
+                        new Warehouse($movement['id_warehouse']),
+                        $quantityToReinject,
+                        null,
+                        $movement['price_te']
+                    );
+                }
+            }
+
+            $productId = $orderDetail->product_id;
+
+            if ($delete) {
+                $orderDetail->delete();
+            }
+
+            StockAvailable::synchronize($productId);
+        } elseif ($orderDetail->id_warehouse == 0) {
+            StockAvailable::updateQuantity(
+                $orderDetail->product_id,
+                $orderDetail->product_attribute_id,
+                $quantityToReinject,
+                $orderDetail->id_shop,
+                true,
+                [
+                    'id_order' => $orderDetail->id_order,
+                    'id_stock_mvt_reason' => Configuration::get('PS_STOCK_CUSTOMER_RETURN_REASON'),
+                ]
+            );
+
+            // sync all stock
+            (new StockManager())->updatePhysicalProductQuantity(
+                (int) $orderDetail->id_shop,
+                (int) Configuration::get('PS_OS_ERROR'),
+                (int) Configuration::get('PS_OS_CANCELED'),
+                null,
+                (int) $orderDetail->id_order
+            );
+
+            if ($delete) {
+                $orderDetail->delete();
+            }
+        } else {
+            throw new OrderException('This product cannot be re-stocked.');
+        }
+    }
+
+    /**
+     * @param Cart $cart
+     * @param Order $order
+     * @param bool $delete
+     *
+     * @return Cart
+     *
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    private function syncCartAndOrderCartRules(Cart $cart, Order $order, bool $delete = false): Cart
+    {
+        Context::getContext()->cart = $cart;
+        CartRule::autoAddToCart();
+        CartRule::autoRemoveFromCart();
+
+        $orderCartRulesData = $order->getCartRules();
+        $orderCartRulesIds = array_map(
+            function (array $orderCartRuleData) { return (int) $orderCartRuleData['id_cart_rule']; },
+            $orderCartRulesData
+        );
+
+        $cartRules = $cart->getCartRules();
+        $cartRulesIds = array_map(
+            function (array $cartRule) { return (int) $cartRule['id_cart_rule']; },
+            $cartRules
+        );
+
+        if (count($orderCartRulesIds) === count($cartRulesIds)) {
```

Comment:
You shouldn't rely on `count` Let's say a cart rule was removed and another one was added the count will be equals but the cart rules are not in sync any more
You can only trust the cart rule id